Library/OcAppleKernelLib/MkextContext.c

```diff
@@ -0,0 +1,1040 @@
+  MkextLength         = SwapBytes32 (Mkext->Header.Length);
+  PlistOffset         = SwapBytes32 (Mkext->PlistOffset);
+  PlistCompressedSize = SwapBytes32 (Mkext->PlistCompressedSize);
+  PlistFullSize       = SwapBytes32 (Mkext->PlistFullSize);
+
+  PlistStoredSize = PlistCompressedSize;
+  if (PlistStoredSize == 0) {
+    PlistStoredSize = PlistFullSize;
+  }
+
+  if (OcOverflowAddU32 (PlistOffset, PlistStoredSize, &Tmp) || Tmp > MkextLength) {
+    return FALSE;
+  }
+
+  PlistBuffer = AllocatePool (PlistFullSize);
+  if (PlistBuffer == NULL) {
+    return FALSE;
+  }
+
+  //
+  // Copy/decompress plist.
+  //
+  if (PlistCompressedSize > 0) {
+    DecompressZLIB (
+      (UINT8 *) PlistBuffer, PlistFullSize,
+      &MkextBuffer[PlistOffset], PlistCompressedSize
+      );
+  } else {
+    CopyMem (PlistBuffer, &MkextBuffer[PlistOffset], PlistFullSize);
+  }
+
+  PlistXml = XmlDocumentParse ((CHAR8 *) PlistBuffer, PlistFullSize, FALSE);
+  if (PlistXml == NULL) {
+    FreePool (PlistBuffer);
+    return FALSE;
+  }
+
+  //
+  // Mkext v2 root element is a dictionary containing an array of bundles.
+  //
+  MkextInfoRoot = PlistNodeCast (XmlDocumentRoot (PlistXml), PLIST_NODE_TYPE_DICT);
+  if (MkextInfoRoot == NULL) {
+    FreePool (PlistBuffer);
+    return FALSE;
+  }
+
+  //
+  // Get bundle array.
+  //
+  MkextInfoRootCount = PlistDictChildren (MkextInfoRoot);
+  for (MkextInfoRootIndex = 0; MkextInfoRootIndex < MkextInfoRootCount; MkextInfoRootIndex++) {
+    BundleArrayKey = PlistKeyValue (PlistDictChild (MkextInfoRoot, MkextInfoRootIndex, &PlistBundleArray));
+    if (AsciiStrCmp (BundleArrayKey, MKEXT_INFO_DICTIONARIES_KEY) == 0) {
+      *Plist        = PlistBuffer;
+      *PlistDoc     = PlistXml;
+      *PlistBundles = PlistBundleArray;
+
+      return TRUE;
+    }
+  }
+
+  //
+  // No bundle array found.
+  //
+  FreePool (PlistBuffer);
+  return FALSE;
+}
+
+STATIC
+UINT32
+UpdateMkextV2Plist (
+  IN MKEXT_V2_HEADER  *Mkext,
+  IN XML_DOCUMENT     *PlistDoc,
+  IN UINT32           Offset
+  )
+{
+  UINT8       *MkextBuffer;
+  CHAR8       *ExportedInfo;
+  UINT32      ExportedInfoSize;
+
+  //
+  // Export plist and include \0 terminator in size.
+  //
+  ExportedInfo = XmlDocumentExport (PlistDoc, &ExportedInfoSize, 0, FALSE);
+  if (ExportedInfo == NULL) {
+    return 0;
+  }
+  ExportedInfoSize++;
+
+  MkextBuffer = (UINT8*)Mkext;
+  CopyMem (&MkextBuffer[Offset], ExportedInfo, ExportedInfoSize);
+  FreePool (ExportedInfo);
+
+  Mkext->PlistOffset = SwapBytes32 (Offset);
+  Mkext->PlistFullSize = SwapBytes32 (ExportedInfoSize);
+  Mkext->PlistCompressedSize = 0;
+  return ExportedInfoSize;
+}
+
+EFI_STATUS
+MkextDecompress (
+  IN     CONST UINT8      *Buffer,
+  IN     UINT32           BufferSize,
+  IN     UINT32           NumReservedKexts,
+  IN OUT UINT8            *OutBuffer,
+  IN     UINT32           OutBufferSize,
+  IN OUT UINT32           *OutMkextSize
+  )
+{
+  BOOLEAN               MkextDecompress;
```

Comment:
Do not make the name of a variable equal to the name of the function. Rename to Decompress for instance.